### ISSUE TITLE
Error: Call to a member function getDefinition() on null

### ISSUE THREAD

**Issue by Eragon666** [sonatra/SonatraSecurityBundle#10]
Hi,

I am experiencing the following error:

CRITICAL - Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalErrorException: "Error: Call to a member function getDefinition() on null" at /home/ubuntu/workspace/vendor/sonatra/security-bundle/Acl/Rule/Definition/AffirmativeDefinition.php line 45 

What I did:
I just added a new user, and I gave the current logged in User the owner permissions. The debugbar shows a 'GRANTED' on the User object. So it seems like the ACL is correct.

Do you know if this is a bug, or how I can fix this?

**Comment by francoispluchino** [comment]
Can you give me your stacktrace?

**Comment by Eragon666** [comment]
Yes, although the stack trace is really short.

```
[1] Symfony\Component\Debug\Exception\FatalErrorException: Error: Call to a member function getDefinition() on null
    at n/a
        in /home/ubuntu/workspace/vendor/sonatra/security-bundle/Acl/Rule/Definition/AffirmativeDefinition.php line 45
```

Debug:
```
INFO - Matched route "{route}". 
DEBUG - Read existing security token from the session. 
DEBUG - SELECT t0.username_canonical AS username_canonical_1, t0.enabled AS enabled_2, t0.salt AS salt_3, t0.last_login AS last_login_4, t0.locked AS locked_5, t0.expired AS expired_6, t0.expires_at AS expires_at_7, t0.confirmation_token AS confirmation_token_8, t0.password_requested_at AS password_requested_at_9, t0.roles AS roles_10, t0.credentials_expired AS credentials_expired_11, t0.credentials_expire_at AS credentials_expire_at_12, t0.id AS id_13, t0.connect_code AS connect_code_14, t0.mumc_id AS mumc_id_15, t0.mumc_AccessToken AS mumc_AccessToken_16, t0.settings AS settings_17, t0.gender AS gender_18, t0.birth AS birth_19, t0.created AS created_20, t0.modified AS modified_21, t0.locale AS locale_22, t0.email_canonical AS email_canonical_23, t0.username AS username_24, t0.email AS email_25, t0.password AS password_26 FROM lcl_user t0 WHERE t0.id = ? LIMIT 1 
DEBUG - User was reloaded from a user provider. 
DEBUG - Notified event "{event}" to listener "{listener}". 
DEBUG - SELECT t0.id AS id_1, t0.name AS name_2, t0.description AS description_3, t0.permissions AS permissions_4, t0.settings AS settings_5, t0.enabled AS enabled_6, t0.created AS created_7, t0.modified AS modified_8 FROM user_role t0 WHERE t0.name IN (?) 
DEBUG - SELECT t0.username_canonical AS username_canonical_1, t0.enabled AS enabled_2, t0.salt AS salt_3, t0.last_login AS last_login_4, t0.locked AS locked_5, t0.expired AS expired_6, t0.expires_at AS expires_at_7, t0.confirmation_token AS confirmation_token_8, t0.password_requested_at AS password_requested_at_9, t0.roles AS roles_10, t0.credentials_expired AS credentials_expired_11, t0.credentials_expire_at AS credentials_expire_at_12, t0.id AS id_13, t0.connect_code AS connect_code_14, t0.mumc_id AS mumc_id_15, t0.mumc_AccessToken AS mumc_AccessToken_16, t0.settings AS settings_17, t0.gender AS gender_18, t0.birth AS birth_19, t0.created AS created_20, t0.modified AS modified_21, t0.locale AS locale_22, t0.email_canonical AS email_canonical_23, t0.username AS username_24, t0.email AS email_25, t0.password AS password_26 FROM lcl_user t0 WHERE t0.id = ? AND (( ( t0.id IN ( SELECT oid.object_identifier FROM acl_entries e JOIN acl_object_identities oid ON ( oid.class_id = e.class_id AND oid.object_identifier != 'class' ) JOIN acl_security_identities s ON ( s.id = e.security_identity_id ) JOIN acl_classes class ON ( class.id = e.class_id ) WHERE e.object_identity_id IS NOT NULL AND (e.mask in (1,4,6,12,16,20,30) OR e.mask >= 32 OR ((e.mask / 2) % 1) > 0) AND oid.id = e.object_identity_id AND (s.identifier = 'UserBundle\\Entity\\User-PintoMatthijs' OR s.identifier = 'ROLE_USER' OR s.identifier = 'ROLE_SYSTEMADMIN' OR s.identifier = 'ROLE_ADMIN') AND class.class_type = 'UserBundle\\Entity\\User' GROUP BY oid.object_identifier)) OR ( SELECT count(oid.object_identifier) FROM acl_entries e JOIN acl_object_identities oid ON ( oid.class_id = e.class_id AND oid.object_identifier = 'class' ) JOIN acl_security_identities s ON ( s.id = e.security_identity_id ) JOIN acl_classes class ON ( class.id = e.class_id ) WHERE e.object_identity_id IS NULL AND (e.mask in (1,4,6,12,16,20,30) OR e.mask >= 32 OR ((e.mask / 2) % 1) > 0) AND (s.identifier = 'UserBundle\\Entity\\User-PintoMatthijs' OR s.identifier = 'ROLE_USER' OR s.identifier = 'ROLE_SYSTEMADMIN' OR s.identifier = 'ROLE_ADMIN' OR s.identifier = 'IS_AUTHENTICATED_FULLY' OR s.identifier = 'IS_AUTHENTICATED_REMEMBERED' OR s.identifier = 'IS_AUTHENTICATED_ANONYMOUSLY') AND class.class_type = 'UserBundle\\Entity\\User' GROUP BY oid.object_identifier >= 1))) LIMIT 1 
DEBUG - SELECT a.ancestor_id FROM acl_object_identities o INNER JOIN acl_classes c ON c.id = o.class_id INNER JOIN acl_object_identity_ancestors a ON a.object_identity_id = o.id WHERE ((o.object_identifier IN ('3','class') AND c.class_type = 'UserBundle\\Entity\\User')) 
DEBUG - SELECT o.id as acl_id, o.object_identifier, o.parent_object_identity_id, o.entries_inheriting, c.class_type, e.id as ace_id, e.object_identity_id, e.field_name, e.ace_order, e.mask, e.granting, e.granting_strategy, e.audit_success, e.audit_failure, s.username, s.identifier as security_identifier FROM acl_object_identities o INNER JOIN acl_classes c ON c.id = o.class_id LEFT JOIN acl_entries e ON ( e.class_id = o.class_id AND (e.object_identity_id = o.id OR e.object_identity_id IS NULL) ) LEFT JOIN acl_security_identities s ON ( s.id = e.security_identity_id ) WHERE (o.id =5 OR o.id = 10) 
DEBUG - ACL found, permission granted. Voting to grant access. 
DEBUG - ACL found, no ACE applicable. Voting to deny access. 
CRITICAL - Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalErrorException: "Error: Call to a member function getDefinition() on null" at /home/ubuntu/workspace/vendor/sonatra/security-bundle/Acl/Rule/Definition/AffirmativeDefinition.php line 45 
```

**Comment by francoispluchino** [comment]
Can you give me your full stacktrace? I don't have enough information.

Otherwise, I thought inject the ACL Rule Manager directly with a DI compiler pass, and not every call of `isGranted` method.

**Comment by Eragon666** [comment]
Where can I find the full stacktrace? Because I have enabled debugmode, and print the full stacktrace (for as far as I know) to my dev.log

This is everything in there about this error:
```
[2016-10-26 21:20:46] request.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalErrorException: "Error: Call to a member function getDefinition() on null" at /home/ubuntu/workspace/vendor/sonatra/security-bundle/Acl/Rule/Definition/AffirmativeDefinition.php line 45 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalErrorException(code: 0): Error: Call to a member function getDefinition() on null at /home/ubuntu/workspace/vendor/sonatra/security-bundle/Acl/Rule/Definition/AffirmativeDefinition.php:45)\n[stacktrace]\n#0 {main}"} []
```

If you want I can give you access to the application for debugging. 

What do you mean with injecting the ACL Rule Manager directly?

**Comment by francoispluchino** [comment]
I will try to push a solution quickly, but I'll make the change tomorrow morning (it is 1:30 AM for me, so, I go to beddy-byes :-)). I hope that it will fix the problem in your configuration.

**Comment by francoispluchino** [comment]
Fixed by 6c07ed5f386f5354cff67d39b19db993e1663ffd. Can you tell me if it's ok for you?

**Comment by Eragon666** [comment]
Yes that seems to work. Thanks for the quick fix, again!
